src/mgr/DaemonServer.cc

```diff
@@ -390,14 +390,11 @@ bool DaemonServer::handle_report(MMgrReport *m)
     dout(20) << "updating existing DaemonState for " << key << dendl;
     daemon = daemon_state.get(key);
   } else {
-    dout(4) << "constructing new DaemonState for " << key << dendl;
-    daemon = std::make_shared<DaemonState>(daemon_state.types);
-    // FIXME: crap, we don't know the hostname at this stage.
-    daemon->key = key;
-    daemon_state.insert(daemon);
-    // FIXME: we should avoid this case by rejecting MMgrReport from
-    // daemons without sessions, and ensuring that session open
-    // always contains metadata.
+    // we don't know the hostname at this stage, reject MMgrReport here.
+    dout(4) << "rejecting report from non-daemon client " << m->daemon_name
+	    << dendl;
+    m->put();
```

Comment:
I don't think it's safe to drop this message -- the MgrClient sessions are stateful, and the client assumes that the server has seen all previous perf counter types it declared (MMgrReport::declare_types).

We need to either delay the message instead of dropping it, or we need to kill the session at this point (hoping that by the time the client re-opens the session, we have created a DaemonState in MetadataUpdate)